app/lib/authenticated_system.rb

```diff
@@ -233,9 +232,9 @@ def send_remember_cookie!
       :expires => @current_user.remember_token_expires_at }
   end
 
-  # Account of current user, or nil if no one is logged in.
+  # This smells of :reek:NilCheck
   def current_account
-    @current_account || (current_user && current_user.account)
+    @current_account ||= current_user&.account&.decorate
```

Comment:
We have a proper setter for the `@current_account` instance variable: https://github.com/3scale/porta/blob/1f077b3237382948e18b88a212978915903c2940/app/lib/authenticated_system.rb#L242-L244

Should we make this a setter as well?

Reply:
I don't understand your proposal 😕

Right now we have the following situation:

- In the UI, we log in by current_user, never current_account. If we have the user, we have the account and we do not need to set it separately IMO. In fact, if we set it separately, we need to ensure that there can never be a mismatch.
- In the API, we log in by current_user (by access token) and by current_account (by provider key). The `current_account` setter is for the latter.
- This method is the reader, not the setter. And it should be 'always' the decorated version (well, we just need it that way right before reaching the views or representers)

Reply:
No proposal actually. I'm just pointing out that this is changing the behaviour.

To agree with you, I see no reason for `current_user` to change along the request (or for the `account` of the current user to change), but current implementation allows that (weirdly or not, who knows?). Your change now makes it to memoize the account, turning this method from a simple getter into a setter. Therefore, a possible new behaviour.

In the end, there's no problem. If the tests are green and you say it's OK doing it, I'm OK with it too. It just seems important to me to highlight when a refactoring change can end up modifying the behaviour too.

Reply:
Maybe it is better to override current_account in a helper and use:

```ruby
module DecoratorHelper
  def current_account
     @_decorated_current_account ||= controller.current_account&.decorate
  end

  def current_user
     @_decorated_current_user ||= controller.current_user&.decorate
  end
```